qiskit_dynamics/models/rotating_frame.py

```diff
@@ -613,54 +605,57 @@ def _enforce_anti_herm(mat: Array, atol: Optional[float] = 1e-10, rtol: Optional
-    mat = Array(mat, dtype=complex)
-
-    if mat.backend == "jax":
-        from jax.lax import cond
-
-        mat = mat.data
-
-        if mat.ndim == 1:
-            # this function checks if pure imaginary. If yes it returns the
-            # array, otherwise it multiplies it by jnp.nan to raise an error
-            # Note: pathways in conditionals in jax cannot raise Exceptions
-            def anti_herm_conditional(b):
-                aherm_pred = jnp.allclose(b, -b.conj(), atol=atol, rtol=rtol)
-                return cond(aherm_pred, lambda A: A, lambda A: jnp.nan * A, b)
-
-            # Check if it is purely real, if not apply anti_herm_conditional
-            herm_pred = jnp.allclose(mat, mat.conj(), atol=atol, rtol=rtol)
-            return Array(cond(herm_pred, lambda A: -1j * A, anti_herm_conditional, mat))
-        else:
-            # this function checks if anti-hermitian, if yes returns the array,
-            # otherwise it multiplies it by jnp.nan
-            def anti_herm_conditional(b):
-                aherm_pred = jnp.allclose(b, -b.conj().transpose(), atol=atol, rtol=rtol)
-                return cond(aherm_pred, lambda A: A, lambda A: jnp.nan * A, b)
-
-            # the following lines check if a is hermitian, otherwise it feeds
-            # it into the anti_herm_conditional
-            herm_pred = jnp.allclose(mat, mat.conj().transpose(), atol=atol, rtol=rtol)
-            return Array(cond(herm_pred, lambda A: -1j * A, anti_herm_conditional, mat))
-
+    mat = unp.asarray(mat)
+
+    try:
+        from jax.lax import cond, convert_element_type
+
+        if isinstance(unp.asarray(mat), jnp.ndarray):
+            if mat.ndim == 1:
+                # this function checks if pure imaginary. If yes it returns the
+                # array, otherwise it multiplies it by jnp.nan to raise an error
+                # Note: pathways in conditionals in jax cannot raise Exceptions
+                def anti_herm_conditional(b):
+                    aherm_pred = jnp.allclose(b, -b.conj(), atol=atol, rtol=rtol)
+                    return convert_element_type(
+                        cond(aherm_pred, lambda A: A, lambda A: jnp.nan * A, b), jnp.complex128
+                    )
+
+                # Check if it is purely real, if not apply anti_herm_conditional
+                herm_pred = jnp.allclose(mat, mat.conj(), atol=atol, rtol=rtol)
+                return unp.asarray(cond(herm_pred, lambda A: -1j * A, anti_herm_conditional, mat))
+            else:
+                # this function checks if anti-hermitian, if yes returns the array,
+                # otherwise it multiplies it by jnp.nan
+                def anti_herm_conditional(b):
+                    aherm_pred = jnp.allclose(b, -b.conj().transpose(), atol=atol, rtol=rtol)
+                    return convert_element_type(
+                        cond(aherm_pred, lambda A: A, lambda A: jnp.nan * A, b), jnp.complex128
+                    )
```

Comment:
Why do we need to use `convert_element_type` here? Is it possible to not reference `jnp.complex128`? Is it possible to do `jnp.complex128` if a user is using 64 bit but `jnp.complex64` if using 32 bit?

Reply:
It is bacause jnp.nan is regarded as float while A is complex.

Reply:
How about is https://github.com/Qiskit-Extensions/qiskit-dynamics/pull/290/commits/8a230f53e4269b014321ad39251b23b2e922f200 ?

Reply:
Adding a suggestion to simplify but it's good otherwise.